### PR TITLE
Add option to  run MIMO input projections on LLM ranks

### PR DESCRIPTION
**Pull request by yashaswikarnati**
## Summary

- optionally place MIMO modality input projections on the first LLM rank with `--mimo-run-input-projections-on-llm-ranks`
- preserve raw bridge payloads while keeping canonical checkpoint keys independent of projector placement
- handle frozen encoders and projector ranks that receive zero local tokens
- avoid tensor truth-value conversion while assembling the global MIMO gradient norm

## Validation

- CW-DFW `16634648` (8 GPUs): the exact models-bucket optimizer fixture regression, frozen-encoder language-rank BF16 optimizer step, and world-consensus test passed on ranks 0-7
- CW-DFW `16634875`: Black and isort checks passed on the final files
- CW-DFW `16633675` (8 GPUs): frozen-encoder language-rank BF16 optimizer step and world-consensus tests passed
- CW-DFW `16627732` (8 GPUs): renamed parser flag, validation, and provider placement tests passed
- CW-DFW `16627995` (8 GPUs): exact 20L checkpoint save/load/resave round trip passed with the renamed flag
- CW-DFW `16616723` (8 GPUs): empty source payload, frozen-encoder language-rank projection, and real zero-token GTP forward/backward tests passed
- CW-DFW `16606559` (8 GPUs): 23 focused parser, construction, freeze/DDP, quantization, empty-payload, and schedule tests passed
- CW-DFW `16606666` (8 GPUs, GTP-enabled): zero-token projector forward/backward and cross-placement checkpoint loading in both directions passed
- CW-DFW `16568629` / `16568652`: functional checkpoint round trips passed with language-rank and encoder-rank projection placement
- GitHub linting, Pip/UV install, wheel build/test, docs, and release checks passed on the preceding head; its models bucket had one stale fake-module fixture failure, fixed and rerun in `16634648`
- `isort --check-only`, `ruff check`, `compileall`, and `git diff --check` passed

### CODE REVIEW

**Review by sancha — APPROVED** [review]

File: megatron/core/models/mimo/model/base.py [hunk 1]
```diff
@@ -138,9 +142,46 @@ def sharded_state_dict(self, prefix='', sharded_offsets=(), metadata=None):
                 while isinstance(inner, (DistributedDataParallel, Float16Module)):
                     inner = inner.module
                     child_prefix += 'module.'
-                sharded_sd.update(
-                    sharded_state_dict_default(inner, child_prefix, sharded_offsets, mod_metadata)
+                module_sd = sharded_state_dict_default(
+                    inner, child_prefix, sharded_offsets, mod_metadata
                 )
+                if (
```
Comment (yashaswikarnati) [inline megatron/core/models/mimo/model/base.py:148]: we need this change so we can still load checkpoints indepedent of if projection layers are used on llm ranks

**Review by yaoyu-33 — APPROVED** [review]

**Review by deepakn94 — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
/ok to test [afc1b15](https://github.com/NVIDIA/Megatron-LM/pull/6843/commits/afc1b15dbe98450e937ea0a7e2681c4e3d86e5ea)

**Comment by yashaswikarnati** [comment]
/ok to test [261ed6f](https://github.com/NVIDIA/Megatron-LM/pull/6843/commits/261ed6fa7b3421990fead75ef27d8455bbb93196)

**Comment by yashaswikarnati** [comment]
/ok to test [e788d04](https://github.com/NVIDIA/Megatron-LM/pull/6843/commits/e788d04c5d8e90bb7d4b29450b3d4c221312b45e)

**Comment by jaredcasper** [comment]
/ok to test [4f02449](https://github.com/NVIDIA/Megatron-LM/pull/6843/commits/4f0244942172f16f26a5d2723b781ef987d7909e)

**Comment by jaredcasper** [comment]
/ok to test [9cd934e](https://github.com/NVIDIA/Megatron-LM/pull/6843/commits/9cd934ee6bb93718f23db7664b1143a535dd93cf)

**Comment by chtruong814** [comment]
Ran CI as part of an aggregated branch that included multiple PRs in the queue.
https://github.com/NVIDIA/Megatron-LM/pull/7459

CI passed
https://github.com/NVIDIA/Megatron-LM/actions/runs/35291473099?pr=7459

Will fast merge this
